buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/ci/metrics/StackdriverMetrics.java

```diff
@@ -0,0 +1,158 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.gradle.plugins.ci.metrics;
+
+import io.opencensus.common.Duration;
+import io.opencensus.exporter.stats.stackdriver.StackdriverStatsConfiguration;
+import io.opencensus.exporter.stats.stackdriver.StackdriverStatsExporter;
+import io.opencensus.stats.Aggregation;
+import io.opencensus.stats.Measure;
+import io.opencensus.stats.Stats;
+import io.opencensus.stats.View;
+import io.opencensus.tags.TagContext;
+import io.opencensus.tags.TagKey;
+import io.opencensus.tags.TagValue;
+import io.opencensus.tags.Tags;
+import io.opencensus.tags.propagation.TagContextBinarySerializer;
+import io.opencensus.tags.propagation.TagContextDeserializationException;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.gradle.api.GradleException;
+import org.gradle.api.Task;
+import org.gradle.api.invocation.Gradle;
+import org.gradle.api.logging.Logger;
+
+/**
+ * Object used to record measurements via {@link #measureSuccess(Task, long)} and {@link
+ * #measureFailure(Task)}.
+ */
+class StackdriverMetrics implements Metrics {
+  private static final AtomicBoolean STACKDRIVER_INITIALIZED = new AtomicBoolean();
+  private static final long STACKDRIVER_UPLOAD_PERIOD_MS = 5000;
+
+  private final TagContext globalContext;
+  private final Logger logger;
+
+  private static Measure.MeasureDouble M_LATENCY =
+      Measure.MeasureDouble.create("latency", "", "ms");
+  private static Measure.MeasureLong M_SUCCESS = Measure.MeasureLong.create("success", "", "1");
+
+  private static final TagKey STAGE = TagKey.create("stage");
+  private static final TagKey GRADLE_PROJECT = TagKey.create("gradle_project");
+
+  private static final List<TagKey> TAG_KEYS =
+      Arrays.asList(
+          STAGE,
+          GRADLE_PROJECT,
+          TagKey.create("repo_owner"),
+          TagKey.create("repo_name"),
+          TagKey.create("pull_number"),
+          TagKey.create("job_name"),
+          TagKey.create("build_id"));
+
+  StackdriverMetrics(Gradle gradle, Logger logger) {
+    this.logger = logger;
+    globalContext = deserializeContext();
+
+    ensureStackdriver(gradle);
+
+    Stats.getViewManager()
+        .registerView(
+            View.create(
+                View.Name.create("fireci/tasklatency"),
+                "The latency in milliseconds",
+                M_LATENCY,
+                Aggregation.LastValue.create(),
+                TAG_KEYS));
+
+    Stats.getViewManager()
+        .registerView(
+            View.create(
+                View.Name.create("fireci/success"),
+                "Indicated success or failure.",
+                M_SUCCESS,
+                Aggregation.LastValue.create(),
+                TAG_KEYS));
+  }
+
+  /** Records failure of the execution stage named {@code name}. */
```

Comment:
Do these methods immediately upload the results? How do you handle errors uploading to Stackdriver?

Reply:
The metrics are batched and uploaded once per the configured `upload interval` which is configured to `5 seconds`. The failures are logged and retried until one of the following: eventual successful upload, process exit.

I think it is fine as metrics are not critical and eventually missing data points is better than to fail the build due to unsuccessful metric upload. wdyt?

Reply:
Nope, it does not throw, it essentially just populates some in-memory data structures